### DISCUSSION TITLE
Passing a large query through a Django template

### DISCUSSION THREAD

**Discussion by numpde** [reactive-python/reactpy-django#118] [Question] [answered]
This may be Django-specific. Let's say I have a view that summons a Django template to a display a long table of widgets, where each widget is an idom component (correponding to a record from a DB table).  As far as I understand, I can pass at most JSONs from a template to a component via `{% component ... %}` -- so each has to refetch its record from the DB if I want to have it (in serial, with `use_query`). Should I `use_context` in the view and the component to share a large query? I feel like I'm missing something...

Thank you.

**Comment by Archmonger** [comment]
If you need "shared data" between several components, you need to create one parent component to contain all your sub-components.

You should use `use_query` within the parent component. This parent component can either provide the query data to the subcomponents via `args`/`kwargs`, or via `use_context`.

**Comment by rmorshea** [comment]
Instead of defining the outer table with a template in Django, it would probably be best to do so using IDOM. This way, you can define an outer component that performs one large query and returns a table of widgets where the data from that query is disseminated to those widgets. As mentioned in [this answer](https://github.com/idom-team/idom/discussions/875#discussioncomment-4636651), you can pass data from that query in two ways - either passing the data down to the components with arguments, or if the components which actually consume the data are sufficiently deep such that passing arguments would be tedious, you can `use_context`. Unfortunately, we don't have documentation for `use_context` yet, but [React's docs](https://beta.reactjs.org/reference/react/useContext#usage) are usually similar enough that you can reference those (especially with respect to best practices and general patterns).

With all that said, here's what that outer table component could look like with and without using a context:

---

**Without context**

```python
from idom import component, html
from django_idom import use_query

@component
def Table():
    data_query = use_query(get_the_data)

    if data_query.loading:
        return html.h2("Loading...")
    elif data_query.error:
        return html.h2("Error!")
    
    return html.table(
        Row(row, key=row.some_unique_id_from_your_data)
        for row in data_query.data
    )

@component
def Row(row_data):
    return html.tr(...)  # render a table row somehow

def get_the_data():
    ...
```

---

**Using context**

```python
from idom import component, html, create_context, use_context
from django_idom import use_query

TableDataContext = create_context(None)

@component
def Table():
    data_query = use_query(get_the_data)

    if data_query.loading:
        return html.h2("Loading...")
    elif data_query.error:
        return html.h2("Error!")
    
    return TableDataContext(
        html.table(
            RowContainerOuter(key=row.some_unique_id_from_your_data)
            for row in data_query.data
        ),
        value=data_query.data
    )

@component
def RowContainerOuter():
    return html.tr(RowContainerInner())

@component
def RowContainerInner():
    return html.tr(Row())

@component
def Row():
    row_data = use_context(TableDataContext)
    return ...  # render a table row somehow

def get_the_data():
    ...
```

**Comment by numpde** [comment]
Thank you both for the answers. Yes, I considered doing it this way, @rmorshea, but I think this is leaving too much Django functionality on the table. 

Anyway, I looked 'under the hood', and my issue seems more related to django-idom==2.1.0. Let me know if I should move it there somehow.

So, I saw that the template tag `component` accepts a `dotted_path` and kwargs. The kwargs are jsonized, fine. The dotted_path is interpreted/imported as a callable, the kwargs are dumped there. I suppose this is the crucial line in `consumer.py` (django_idom):

```Python
component_instance = component_constructor(**component_kwargs)
```

I'd like to have an option to pass the Django `context` also. This seems very natural. I probably wouldn't change the above part, but redefine the callable `component_constructor` by binding `context` to the first argument, say, before it lands in `IDOM_REGISTERED_COMPONENTS`. So, perhaps inside the `component` tag definition, this line

```Python
_register_component(dotted_path)
```

would be

```Python
_register_component(context, dotted_path)
```

and therein, 

```Python
IDOM_REGISTERED_COMPONENTS[dotted_path] = partial(_import_dotted_path(dotted_path), context)
```

What do you think?

Since I can pass anything to an idom component, I don't see why it shouldn't be possible for a template tag.

**Reply by Archmonger** [reply]
> . . . but I think this is leaving too much Django functionality on the table.
> . . .
> Since I can pass anything to an idom component, I don't see why it shouldn't be possible for a template tag.

I agree with your sentiment, and it would be nice to engineer a way to pass Django template context into the top-level components.

However, pushing the Django context into `IDOM_REGISTERED_COMPONENTS` wouldn't work. Contexts can/will have session/user data, and unfortunately the `IDOM_REGISTERED_COMPONENTS` is designed to be a global/non-scoped/generic list of components.

Off the top of my head, a way I can see this implementation working is having it be database-backed.

The template tag could [pickle](https://pypi.org/project/dill/) and store the context in the database. The database model would use the component `uuid` as the `primary_key` or alternatively use the [`session`](https://docs.djangoproject.com/en/4.1/topics/http/sessions/) as a `ForiegnKey`.

When it comes to retrieving this data within a component, I could imagine a `use_django_context` hook that relies on [`use_query`](https://github.com/idom-team/django-idom/blob/8a7b64195009206f93f9bb4de4e8e07d1d7feb45/src/django_idom/hooks.py#L102) to fetch the data from the database.

---

Taking this a step further, the proposed implementation above could instead pickle `component_kwargs` (rather than `context`), which would allow for passing in quite literally anything into our template tag.

I believe this would solve the same issue, but not require a `use_django_context` hook.

**Reply by rmorshea** [reply]
Could we just pickle the kwargs passed to the component instead of serializing them as JSON?

**Reply by Archmonger** [reply]
Yup, we had the same thought. I edited my comment to include that right before I saw your reply.

**Reply by rmorshea** [reply]
Doing this comes at the cost of growing the size of the template (since that's where the pickled data ends up), but the benefit of course is simplicity and transparency for the user. My intuition says that, if we document this fact (i.e. that you should only pass exactly the data you need via the template) that this approach of pickling kwargs into the template should work in a majority of cases.

**Reply by Archmonger** [reply]
Doing this would also cause a performance penalty though. Every component render would require 2 database queries (`set` and `get`).

**Reply by Archmonger** [reply]
Nope, it would technically reduce template size. Component kwargs would get plucked away from the HTML, and instead be commited into the database during the HTTP response cycle, and fetched via our websocket consumer using the component `uuid`.

**Reply by Archmonger** [reply]
Ohhhh I see, you're saying just pickle the whole object and plop it onto the page, rather than the database....

Hmmm... that sounds like it has some scary ramifications. If the user modifies his HTML and we try to unpickle that object, that would become a remote-code execution attack vector. Would be fairly trivial to override things like `__init__` to perform any various malicious task.

I don't believe it's viable to store any Python objects client sided.

**Reply by rmorshea** [reply]
Oh yeah, that's a bad idea 😨

**Reply by Archmonger** [reply]
Thinking about it now, using database-backed `component_kwargs` might be mandatory.

If the component has a kwarg called `is_admin`, it's currently fairly trivial to edit the page HTML and set that to `true` within the JSON parameters.

**Reply by rmorshea** [reply]
Agreed. Too much room for error there.

**Reply by Archmonger** [reply]
I will PR this ASAP.

**Reply by rmorshea** [reply]
Before making a PR, can you lay out the approach in an issue? I just want to be sure we hash out all the subtleties. For example, is this data transient? If so, how do we clean up the data when it's no longer needed. If not, then unpickling previously stored objects could be an issue if users upgrade their Python version.

**Reply by Archmonger** [reply]
- #119

**Comment by Archmonger** [comment, marked as answer]
This limitation will be resolved in v3.0.0

Template tag parameters will now have unrestricted `arg`/`kwarg` values. Thus, they will perform identically to a typical Django template tag.

See the following PR for more details:

- #120

**Reply by numpde** [reply]
Fwiw, I think this is great (and having tested my proposal above: it was crap). 

I wonder about two things:
- Will {% component %} work as expected if invoked with different parameters from within a {% for %} loop in a template? In other words, what is the default 'key'?
- What changes to the objects passed as arguments will trigger an update of the component (I suppose none by default)?

**Reply by Archmonger** [reply]
- **Will {% component %} work as expected if invoked with different parameters from within a {% for %} loop in a template? In other words, what is the default 'key'?**
   - Yes, each individual component is it's own unique ReactJS DOM tree. Therefore, even if the `key` of the top level node is identical it won't matter.
- **What changes to the objects passed as arguments will trigger an update of the component (I suppose none by default)?**
   - Any time the Django template is re-rendered, the components will be re-rendered as well.
